src/simd.rs

```diff
@@ -0,0 +1,215 @@
+//! SIMD byte scanning logic.
+//! 
+//! This module provides functions that allow walking through byteslices, calling
+//! provided callback functions on special bytes and their indices using SIMD.
+//! The byteset is defined in `compute_lookup`.
+//!
+//! The idea is to load in a chunk of 16 bytes and perform a lookup into a set of
+//! bytes on all the bytes in this chunk simultaneously. We produce a 16 bit bitmask
+//! from this and call the callback on every index corresponding to a 1 in this mask
+//! before moving on to the next chunk. This allows us to move quickly when there
+//! are no or few matches.
+//!
+//! The table lookup is inspired by this [great overview]. However, since all of the
+//! bytes we're interested in are ASCII, we don't quite need the fully generality of
+//! the universal algorithm and are hence able to skip a few instructions.
+//!
+//! [great overview]: http://0x80.pl/articles/simd-byte-lookup.html
+
+use core::arch::x86_64::*;
+use crate::parse::LoopInstruction;
+
+const VECTOR_SIZE: usize = std::mem::size_of::<__m128i>();
+
+/// Generates a lookup table containing the bitmaps for our
+/// special marker bytes. This is effectively a 128 element 2d bitvector,
+/// that can be indexed by a four bit row index (the lower nibble)
+/// and a three bit column index (upper nibble).
+const fn compute_lookup() -> [u8; 16] {
+    let mut lookup = [0u8; 16];
+    lookup[(b'\n' & 0x0f) as usize] |= 1 << (b'\n' >> 4);
+    lookup[(b'\r' & 0x0f) as usize] |= 1 << (b'\r' >> 4);
+    lookup[(b'*'  & 0x0f) as usize] |= 1 << (b'*'  >> 4);
+    lookup[(b'_'  & 0x0f) as usize] |= 1 << (b'_'  >> 4);
+    lookup[(b'~'  & 0x0f) as usize] |= 1 << (b'~'  >> 4);
+    lookup[(b'|'  & 0x0f) as usize] |= 1 << (b'|'  >> 4);
+    lookup[(b'&'  & 0x0f) as usize] |= 1 << (b'&'  >> 4);
+    lookup[(b'\\' & 0x0f) as usize] |= 1 << (b'\\' >> 4);
+    lookup[(b'['  & 0x0f) as usize] |= 1 << (b'['  >> 4);
+    lookup[(b']'  & 0x0f) as usize] |= 1 << (b']'  >> 4);
+    lookup[(b'<'  & 0x0f) as usize] |= 1 << (b'<'  >> 4);
+    lookup[(b'!'  & 0x0f) as usize] |= 1 << (b'!'  >> 4);
+    lookup[(b'`'  & 0x0f) as usize] |= 1 << (b'`'  >> 4);
+    lookup
+}
+
+/// Computes a bit mask for the given byteslice starting from the given index,
+/// where the 16 least significant bits indicate (by value of 1) whether or not
+/// there is a special character at that byte position. The least significant bit
+/// corresponds to `bytes[ix]` and the most significant bit corresponds to
+/// `bytes[ix + 15]`.
+/// It is only safe to call this function when `bytes.len() >= ix + VECTOR_SIZE`.
+unsafe fn compute_mask(bytes: &[u8], ix: usize) -> i32 {
```

Comment:
This needs a `#[target_feature(enable = "ssse3")]`, otherwise you might get less ideal code generation. It might also be worth adding an `#inline` - you'd need to look at generated asm to be very confident this is being done right. (If the generated asm is already good, it's because the compiler is automatically determining that inlining is appropriate, and the `target_feature_enable` on `simd_iterate_special_bytes` is in effect. But that's a bit fragile, for example if it's compiled optimized for size the inlining decision might go another way).